docs/docs/features/soft-off.md

```diff
@@ -0,0 +1,164 @@
+
+#### GPIO Key
+
+Zephyr's basic GPIO Key concept is used as a shared configuration of the details of the GPIO pin that will be used both for triggering soft off, and for waking the device latter. Here is an example for a keyboard with a dedicated on/off push button that is a direct wire between the GPIO pin, and ground:
+
+```
+/ {
+    keys {
+        compatible = "gpio-keys";
+        wakeup_key: wakeup_key {
+            gpios = <&gpio0 2 (GPIO_ACTIVE_LOW | GPIO_PULL_UP)>;
+        };
+    };
+};
+```
+
+GPIO keys are defined by using child nodes under the `gpio-keys` compatible node. Each child needs just one property defined:
+
+- The `gpios` property should be a phandle-array with a fully defined GPIO pin and with the correct pull up/down and active high/low flags set. In the above example the soft on/off would be triggered by pulling the pin low by pressing a switch that has the other leg connected to ground.
+
+#### Behavior Key
+
+Next, we will create a new "behavior key". Behavior keys are an easy way to tie a keymap behavior to a GPIO key outside of the normal keymap processing. They do _not_ do the normal keymap processing, so they are only suitable for use with basic behaviors, not complicated macros, hold-taps, etc.
+
+In this case, we will be creating a dedicated instance of the [Soft Off Behavior](../behaviors/soft-off.md) that will be used only for our hardware on/off button, then binding it to our key:
+
+```
+/ {
+    behaviors {
+        hw_soft_off: behavior_hw_soft_off {
+            compatible = "zmk,behavior-soft-off";
+            #binding-cells = <0>;
+            label = "HW_SO";
+            hold-time-ms = <5000>;
+        };
+    };
+
+    soft_off_behavior_key {
+        compatible = "zmk,behavior-key";
+        bindings = <&hw_soft_off>;
+        key = <&wakeup_key>;
+    };
+};
+```
+
+Here are the properties for the behavior key node:
+
+- The `compatible` property for the node must be `zmk,behavior-key`.
+- The `bindings` property is a phandle to the soft off behavior defined above.
+- The `key` property is a phandle to the GPIO key defined earlier.
+
+If you have set up your on/off to be controlled by a matrix integrated combo, the behavior key needs use a different driver that will handle detecting the pressed state when the pin is toggled by the other matrix kscan driver:
+
+```
+/ {
+    soft_off_behavior_key {
+        compatible = "zmk,behavior-key-scanned";
+        status = "okay";
+        bindings = <&hw_soft_off>;
+        key = <&wakeup_key>;
+    };
+};
+```
+
+The only difference is using a `compatible` value of `zmk,behavior-key-scanned`.
+
+#### Wakeup Sources
+
+Zephyr has general support for the concept of a device as a "wakeup source", which ZMK has not previously used. Adding soft off requires properly updating the existing `kscan` devices with the `wakeup-source` property, e.g.:
+
+```
```

Comment:
I believe it's necessary to be sure that it *doesn't* get shut down when doing an automatic deep sleep, now that device PM is enabled. I will double check this all to be sure though.

Reply:
Yeah, exactly this. There's added code to `src/kscan.c` that will enable the device as a wakeup (tagging in DTS just marks it as being wakeup source *capable*, not that it's enabled as a wakeup source), so that suspend with device PM enabled will keep the kscan device enabled, so it'll be able to wake the device from "normal" deep sleep. This piece is all about keeping normal deep sleep happy.